### ISSUE TITLE
Some "should not happen" panics

### ISSUE THREAD

**Issue by bep** [go-yaml/yaml#116]
```
package main

import (
    "fmt"
    "gopkg.in/yaml.v2"
)

func main() {

    data := [][]byte{[]byte("---\n- - - - - - - - " +
        "{\n0}0\n---"), []byte("---\n ? \"\n\"0\n---"), []byte("---\n- - - - - - {\n0}" +
        "0\n---"), []byte("---\n- - - - - - - - " +
        "{\n0}0\n---"), []byte("---\n0:\n ? \"\n\"0\n---")}

    for _, d := range data {
        doTest(d)
    }
}

func doTest(d []byte) {
    defer func() {
        if r := recover(); r != nil {
            fmt.Println("Recovered panic: ", r)
        }

    }()
    m := map[string]interface{}{}
    yaml.Unmarshal(d, &m)
}
```

The variations above, and some other similar, panics with the message "should not happen".
If the data above looks constructed, that's because it is: https://github.com/dvyukov/go-fuzz

**Comment by asib** [comment]
Found a couple more myself

``` go
package main

import (
    "fmt"

    "gopkg.in/yaml.v2"
)

func main() {
    data := [][]byte{[]byte("0: \n  0: [0\n0]0"),
        []byte("  0: [\n0]0")}

    for _, d := range data {
        doTest(d)
    }
}

func doTest(d []byte) {
    defer func() {
        if r := recover(); r != nil {
            fmt.Println("recovered panic: ", r)
        }
    }()

    yaml.Unmarshal(d, &struct{}{})
}
```

**Comment by jorinvo** [comment]
I was also running go-fuzzer and I can confirm that the above errors still exist.

Additionally I found two index out of range errors:

```
"{\n0: \"00\\\r\n",
"0: \"0\\\r\n",
"{\"0\\\r\n",
```

These strings alle cause the following error:

```
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 1 [running]:
gopkg.in/yaml%2ev2.handleErr(0xc4200fbbe8)
	/var/folders/km/kqlfxy2j64dd_9gqb373c0hm0000gn/T/go-fuzz-build867195819/gopath/src/gopkg.in/yaml.v2/yaml.go:153 +0xf7
panic(0x14743a0, 0x1706050)
	/var/folders/km/kqlfxy2j64dd_9gqb373c0hm0000gn/T/go-fuzz-build867195819/goroot/src/runtime/panic.go:489 +0x2cf
gopkg.in/yaml%2ev2.yaml_parser_scan_flow_scalar(0xc4200e8840, 0xc4200fb388, 0x0, 0x1)
	/var/folders/km/kqlfxy2j64dd_9gqb373c0hm0000gn/T/go-fuzz-build867195819/gopath/src/gopkg.in/yaml.v2/scannerc.go:2483 +0x2c1d
gopkg.in/yaml%2ev2.yaml_parser_fetch_flow_scalar(0xc4200e8840, 0x0, 0x1)
	/var/folders/km/kqlfxy2j64dd_9gqb373c0hm0000gn/T/go-fuzz-build867195819/gopath/src/gopkg.in/yaml.v2/scannerc.go:1407 +0xe2
```

This string causes a different out of range error:
```
"\u0085\ufeff\u0085"
```

```
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 1 [running]:
gopkg.in/yaml%2ev2.handleErr(0xc42004bbe8)
	/var/folders/km/kqlfxy2j64dd_9gqb373c0hm0000gn/T/go-fuzz-build867195819/gopath/src/gopkg.in/yaml.v2/yaml.go:153 +0xf7
panic(0x14743a0, 0x1706050)
	/var/folders/km/kqlfxy2j64dd_9gqb373c0hm0000gn/T/go-fuzz-build867195819/goroot/src/runtime/panic.go:489 +0x2cf
gopkg.in/yaml%2ev2.yaml_parser_scan_to_next_token(0xc4204242c0, 0xc4204242c0)
	/var/folders/km/kqlfxy2j64dd_9gqb373c0hm0000gn/T/go-fuzz-build867195819/gopath/src/gopkg.in/yaml.v2/scannerc.go:1455 +0x4e0
gopkg.in/yaml%2ev2.yaml_parser_fetch_next_token(0xc4204242c0, 0x1)
	/var/folders/km/kqlfxy2j64dd_9gqb373c0hm0000gn/T/go-fuzz-build867195819/gopath/src/gopkg.in/yaml.v2/scannerc.go:677 +0xc5
```

**Comment by niemeyer** [comment]
Some related issues were fixed with #323, and I've confirmed that none of these panic anymore.
